### PR TITLE
Migration to Python 3 (also compatible with Python 2)

### PR DESCRIPTION
**Pull request by qtheuret**
Run 2to3 script on code and make some tests with Python 2.7 and Python 3.6.5. It works like a charm.

I ran tests with Prestashop 1.7 but I didn't change the MAX_COMPATIBLE_VERSION value because I didn't make all possible cases. I'll do a new merge request to increment the MAX_COMPATIBLE_VERSION when they will be done.

### CODE REVIEW

File: examples/prestapyt_dict.py [hunk 1]
```diff
@@ -2,8 +2,8 @@
 from xml.etree import ElementTree
 
 
-prestashop = PrestaShopWebService('http://localhost:8080/api',
-                                  'BVWPFFYBT97WKM959D7AVVD0M4815Y1L',
+prestashop = PrestaShopWebService('http://localhost/api',
+                                  'HRT7FAHK3HPQDXEYG5DF1BDXDUJ61MIR',
```
Comment (pedrobaeza) [inline examples/prestapyt_dict.py]: Why this change?

File: pypi_commands.sh [hunk 2]
```diff
@@ -1,4 +1,4 @@
 #! /bin/sh
-python2.7 setup.py sdist upload
```
Comment (pedrobaeza) [inline pypi_commands.sh]: We should deploy to both versions

**Comment by qtheuret** [comment]
@pedrobaeza I revert the API key and add command to deploy on both versions.

**Comment by capiscuas** [comment]
Would be great if this can be merged ;-) I am in the same situation with python3.5

File: prestapyt/dict2xml.py [hunk 3]
```diff
@@ -21,15 +21,15 @@ def _process(doc, tag, tag_value):
...
+    if isinstance(tag_value, dict) and 'value' in list(tag_value.keys()) == ['value']:
         tag_value = tag_value['value']
 
     if tag_value is None:
         tag_value = ''
 
     # Create a new node for simple values
     if (isinstance(tag_value, (float, int)) or
-            isinstance(tag_value, (str, unicode))):
+            isinstance(tag_value, str)):
```
Comment (guewen) [inline prestapyt/dict2xml.py]: This will break compatibility with python2, you can add the `future` library and use this pattern:

```
from past.builtins import basestring

... isinstance(tag_value, basestring)
```

File: prestapyt/dict2xml.py [hunk 4]
```diff
@@ -104,7 +104,7 @@ def _process_simple(doc, tag, tag_value):
     @return: node
     """
     node = doc.createElement(tag)
-    node.appendChild(doc.createTextNode(unicode(tag_value)))
+    node.appendChild(doc.createTextNode(str(tag_value)))
```
Comment (guewen) [inline prestapyt/dict2xml.py]: same remark with py2/py3 different behavior

File: prestapyt/prestapyt.py [hunk 5]
```diff
@@ -42,9 +45,10 @@
 try:  # for Python 3
     from http.client import HTTPConnection
 except ImportError:
-    from httplib import HTTPConnection
+    from http.client import HTTPConnection
```
Comment (guewen) [inline prestapyt/prestapyt.py]: Nope, look at the content of the try except, this is specifically the python 2 import.
Note that normally `from http.client import HTTPConnection` should work even with py2 if python-futures is installed.

File: prestapyt/prestapyt.py [hunk 6]
```diff
@@ -319,7 +323,7 @@ def _options_to_querystring(self, options):
         """
         if self.debug:
             options.update({'debug': True})
-        return urllib.urlencode(options)
+        return urllib.parse.urlencode(options)
```
Comment (guewen) [inline prestapyt/prestapyt.py]: Won't work with both py2 and py3: http://python-future.org/compatible_idioms.html#urllib-module

File: prestapyt/xml2dict.py [hunk 7]
```diff
@@ -14,14 +14,14 @@
...
-except ImportError, err:
+except ImportError as err:
     import xml.etree.ElementTree as ET
 
 
 def _parse_node(node):
     tree = {}
     attrs = {}
-    for attr_tag, attr_value in node.attrib.items():
+    for attr_tag, attr_value in list(node.attrib.items()):
```
Comment (guewen) [inline prestapyt/xml2dict.py]: `list` useless here

**Comment by guewen** [comment]
> What is your position on let this code only for Python 3 or make changes to do it compatible 2/3 ?

No strong feeling on this. Actually, I don't really use and then work on the lib since years.
@pedrobaeza, @bguillot your opinion?

**Comment by pedrobaeza** [comment]
I prefer to keep compatibility with both versions for now, as the same lib is used on Odoo 9.0/10.0 deployments.

File: examples/prestapyt_xml.py [hunk 8]
```diff
@@ -8,17 +8,17 @@
 
 prestashop.get('')
 
-print "GET"
-print ElementTree.tostring(prestashop.get('addresses'))
-print ElementTree.tostring(prestashop.get('addresses', resource_id=1))
-print ElementTree.tostring(prestashop.get('addresses/1'))
-print ElementTree.tostring(prestashop.get('addresses', options={'limit': 1}))
-print ElementTree.tostring(prestashop.get('stock_movement_reasons'))
+print("GET")
```
Comment (pedrobaeza) [inline examples/prestapyt_xml.py]: Better to create a new file with `_python3` suffix or similar and let this one for those using Python 2.

**Comment by alu-** [comment]
Can we get this merged?

**Comment by alu-** [comment]
@qtheuret are you working on the changes requested by @pedrobaeza ?

File: examples/prestapyt_xml.py (hunk 8 above)
Comment (qtheuret) [reply examples/prestapyt_xml.py]: For example files only or for the whole library ?

**Comment by pedrobaeza** [comment]
The library should be compatible with both Python 2/3 (with the same code), and there should be an example file in Python 2 and another in Python 3 (or use the same but compatible with both versions. You can use `from future import print_function` for example.

**Comment by qtheuret** [comment]
Requested changed done.

**Comment by kelu27** [comment]
Can we get this merged? @pedrobaeza

**Comment by FFernandez-PlanetaTIC** [comment]
Yes, with this library version these tests pass:

- v10.0 connector_prestashop (OCA/connector-prestashop#108)
- v9.0 connector_prestashop_catalog_manager (OCA/connector-prestashop#99)

**Comment by pedrobaeza** [comment]
@FFernandez-PlanetaTIC just a last confirmation before merging. Did you talk about tests, but tests are mocked if I'm right. Have you tried in a real environment?
